### PR TITLE
Implement pagination, order by and groups in search

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/handlers/search.py [hunk 1]
```diff
@@ -77,19 +120,131 @@ def search(self, user, content):
 
         room_ids = search_filter.filter_rooms(room_ids)
 
-        rank_map, event_map, _ = yield self.store.search_msgs(
-            room_ids, search_term, keys
-        )
+        if batch_group == "room_id":
+            room_ids = room_ids & {batch_group_key}
```
Comment (NegativeMjark) [inline synapse/handlers/search.py]: Why not ``room_ids.add(batch_group_key)``?

Comment (NegativeMjark) [inline synapse/handlers/search.py]: Ah. Cause `&` mean intersection.

Comment (NegativeMjark) [inline synapse/handlers/search.py]: Maybe ``room_ids.intersection_update(batch_group_key)`` ?

File: synapse/handlers/search.py [hunk 2]
```diff
@@ -77,19 +120,131 @@ def search(self, user, content):
...
+                    pagination_token = batch_token
+                else:
+                    pagination_token = None
+                i = 0
+
+                # We keep looping and we keep filtering until we reach the limit
+                # or we run out of things.
+                # But only go around 5 times since otherwise synapse will be sad.
+                while len(room_events) < search_filter.limit() and i < 5:
+                    i += 5
```
Comment (NegativeMjark) [inline synapse/handlers/search.py]: I'm not sure that this loops 5 times if ``i`` is being incremented by 5.

File: synapse/handlers/search.py [hunk 3]
```diff
@@ -77,19 +120,131 @@ def search(self, user, content):
...
+                        pagination_token = None
+                        break
+                    else:
+                        pagination_token = results[-1]["pagination_token"]
+
+                if room_events:
+                    res = results_map[room_events[-1].event_id]
+                    pagination_token = res["pagination_token"]
+
+                if room_events:
```
Comment (NegativeMjark) [inline synapse/handlers/search.py]: This test seems a little bit redundant given the test about.

File: synapse/handlers/search.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/handlers/search.py]: Not `room_ids &= batch_group_key` ? :p

File: synapse/handlers/search.py (hunk 2 above)
Comment (erikjohnston) [inline synapse/handlers/search.py]: Ha

**Comment by NegativeMjark** [comment]
LGTM modulo not really knowing what the API is supposed to look like.

**Pull request merged**